server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java

```diff
@@ -413,10 +423,21 @@ List<Diagnosis.Definition> diagnoseUnassignedShardRouting(ShardRouting shardRout
                     actions.add(ACTION_RESTORE_FROM_SNAPSHOT);
                 }
                 break;
+            case NO_ATTEMPT:
+                if (shardRouting.unassignedInfo().isDelayed()) {
+                    actions.add(DIAGNOSIS_WAIT_FOR_OR_FIX_DELAYED_SHARDS);
+                } else {
+                    actions.addAll(explainAllocationsAndDiagnoseDeciders(shardRouting, state));
+                }
+                break;
             case DECIDERS_NO:
                 actions.addAll(explainAllocationsAndDiagnoseDeciders(shardRouting, state));
                 break;
+            case DELAYED_ALLOCATION:
+                actions.add(DIAGNOSIS_WAIT_FOR_OR_FIX_DELAYED_SHARDS);
+                break;
             default:
+                actions.addAll(explainAllocationsAndDiagnoseDeciders(shardRouting, state));
```

Comment:
Do we need to expand the integration tests for this change? (or pull it into a separate PR and do it as part of that?)

Reply:
Alright so I really tried to get a test case where this default case is exercised but I can't manage it. 

The only remaining `lastAllocationStatus` values possible are `FETCHING_SHARD_DATA` and `DECIDERS_THROTTLED`. After some digging, it looks like these are only used for throttled recoveries and for searchable snapshots that have not finished reporting their stats, neither of which we currently provide debugging information for. Additionally, these states are almost always transient, and attempting to lodge them into the state for extended periods (like creating an index while setting concurrent recoveries to 0) has the health API skipping the diagnosis logic anyway (treating indices as new and thus not important to report failures for yet, timely shutdowns, etc). I don't think we even have logic to detect problems pertaining to these unassigned shard cases right now. Should we remove the default here?

Reply:
Thanks for looking into this Jimmy.

I played around with this quite a bit today and I think `NO_ATTEMPT` covers some of the cases I had in mind.

e.g.

```
    public void testIndexDisabledAllocation() throws Exception {
        internalCluster().startMasterOnlyNodes(1);
        internalCluster().startNodes(1, onlyRole(DiscoveryNodeRole.DATA_HOT_NODE_ROLE));
        ElasticsearchAssertions.assertAcked(
            prepareCreate("test").setSettings(
                Settings.builder()
                    .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
                    .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
                    .put(DataTier.TIER_PREFERENCE, DataTier.DATA_HOT)
            )
        );
        ensureGreen("test");

        Settings.Builder settingsBuilder = Settings.builder()
            .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 22)
            .put(EnableAllocationDecider.INDEX_ROUTING_ALLOCATION_ENABLE_SETTING.getKey(), "none");
        UpdateSettingsRequestBuilder settingsRequest = client().admin().indices().prepareUpdateSettings("test");
        settingsRequest.setSettings(settingsBuilder);
        settingsRequest.setTimeout("1s").setMasterNodeTimeout("1s");
        settingsRequest.execute().actionGet();

        ensureYellow("test");

        GetHealthAction.Response healthResponse = client().execute(
            GetHealthAction.INSTANCE,
            new GetHealthAction.Request(ShardsAvailabilityHealthIndicatorService.NAME, true)
        ).get();
        HealthIndicatorResult indicatorResult = healthResponse.findIndicator(ShardsAvailabilityHealthIndicatorService.NAME);
        assertThat(indicatorResult.status(), equalTo(HealthStatus.YELLOW));
        assertThat(
            indicatorResult.diagnosisList(),
            hasItem(
                new Diagnosis(
                    ShardsAvailabilityHealthIndicatorService.ACTION_ENABLE_INDEX_ROUTING_ALLOCATION,
                    List.of("test")
                )
            )
        );
    }
```

>Should we remove the default here?

Yeah let's do that.

Toying around with the integratino tests I stumbled a bit when grepping for the class. Should we rename `DataTierShardAvailabilityHealthIndicatorIT` to `ShardAvailabilityHealthIndicatorIT` ? 
Or do you envision tests that don't have to do with tier-ing (like the one I posted above) to be in a future `ShardAvailabilityHealthIndicatorIT` ?

Reply:
I think `DataTierShardAvailabilityHealthIndicatorIT` is appropriate for now since that test requires the data tiers plugin to be installed to operate. I had originally called it just `ShardAvailabilityHealthIndicatorIT` when I started building it in the server project, but since almost all of the test cases of the health indicator depend on data tiers code I moved it and renamed it thusly.